### PR TITLE
fix(core): default to whole file change type

### PR DESCRIPTION
**Pull request by FrozenPandaz**
## Current Behavior (This is the behavior we have today, before the PR is merged)

`yarn affected:apps` (no other args) errors with:

```sh
 jason@pop-os  ~/projects/nx-example   master ●  yarn affected:apps                             
yarn run v1.19.0
$ nx affected:apps
/home/jason/projects/nx-example/node_modules/yargs/yargs.js:1109
      else throw err
           ^

TypeError: Cannot read property 'length' of undefined
    at Object.parseFiles (/home/jason/projects/nx-example/node_modules/@nrwl/workspace/src/command-line/shared.js:64:24)
    at readFileChanges (/home/jason/projects/nx-example/node_modules/@nrwl/workspace/src/command-line/affected.js:94:28)
    at Object.affected (/home/jason/projects/nx-example/node_modules/@nrwl/workspace/src/command-line/affected.js:17:25)
    at Object.exports.commandsObject.yargs.usage.command.command.command.command.command.args [as handler] (/home/jason/projects/nx-example/node_modules/@nrwl/workspace/src/command-line/nx-commands.js:63:113)
    at Object.runCommand (/home/jason/projects/nx-example/node_modules/yargs/lib/command.js:235:44)
    at Object.parseArgs [as _parseArgs] (/home/jason/projects/nx-example/node_modules/yargs/yargs.js:1022:30)
    at Object.get [as argv] (/home/jason/projects/nx-example/node_modules/yargs/yargs.js:965:21)
    at Object.initLocal (/home/jason/projects/nx-example/node_modules/@nrwl/cli/lib/init-local.js:18:26)
    at Object.<anonymous> (/home/jason/projects/nx-example/node_modules/@nrwl/cli/bin/nx.js:12:18)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

JSON diff is also not calculated properly

## Expected Behavior (This is the new behavior we can expect after the PR is merged)

`yarn affected:apps` (no other args) defaults to `--base master --head HEAD` and JSON diff is calculated properly.

### CODE REVIEW

File: packages/workspace/src/core/file-utils.ts [hunk 1]
```diff
@@ -9,6 +9,7 @@ import { readJsonFile } from '../utils/fileutils';
 import { Environment, NxJson } from './shared-interfaces';
 import { ProjectGraphNode } from './project-graph';
 import { WorkspaceResults } from '../command-line/workspace-results';
+import { NxArgs } from '@nrwl/workspace/src/command-line/utils';
```
Comment (jaysoo) [inline packages/workspace/src/core/file-utils.ts]: Use relative import here.

**Review by jaysoo — COMMENTED** [review]
I like this idea. I wonder how this affects downstream work. 

For example, for implicit JSON dependencies, we'll need to make a special case to handle `WholeFileModification` correct? Same for other processors.

File: packages/workspace/src/core/file-utils.ts [hunk 2]
```diff
@@ -35,29 +44,30 @@ export function calculateFileChanges(
...
-                const atHead = readFileAtRevision(f, head);
-                value = jsonDiff(atBase, atHead);
-              } else {
-                value = [];
-              }
-              break;
-            default:
-              throw new Error(`Cannot call getChanges() on ${f}`);
-          }
+      getChanges: (): Modification[] => {
```
Comment (vsavkin) [inline packages/workspace/src/core/file-utils.ts]: See the comment above. getChanges should return an array of changes, not modifications.

File: packages/workspace/src/core/file-utils.ts [hunk 3]
```diff
@@ -18,14 +19,22 @@ export interface FileData {
   ext: string;
 }
 
-export interface FileChange extends FileData {
-  getChanges: () => any[];
+export interface Modification {
```
Comment (vsavkin) [inline packages/workspace/src/core/file-utils.ts]: I think we should either use the word modifictaion or the word change. I prefer change.

File: packages/workspace/src/core/file-utils.ts [hunk 4]
```diff
@@ -35,29 +44,30 @@ export function calculateFileChanges(
...
-          }
+      getChanges: (): Modification[] => {
+        if (!nxArgs) {
+          return [WholeFileChange];
+        }
+        switch (ext) {
+          case '.json':
+            const atBase = nxArgs.base
+              ? readFileAtRevision(f, nxArgs.base)
+              : readFileAtRevision(f, 'master');
```
Comment (vsavkin) [inline packages/workspace/src/core/file-utils.ts]: should not hardcode master here. We should default base to master.

File: packages/workspace/src/command-line/affected.ts [hunk 5]
```diff
@@ -113,14 +117,8 @@ export function affected(command: string, parsedArgs: yargs.Arguments): void {
...
 
-function readFileChanges(nxArgs: NxArgs) {
-  // Do we still need this `--all` option?
-  if (nxArgs.all) {
-    return [];
-  }
-
-  const files = parseFiles(nxArgs).files;
-  return calculateFileChanges(files, nxArgs.base, nxArgs.head);
+function readFileChanges(files: string[], nxArgs: NxArgs) {
```
Comment (vsavkin) [inline packages/workspace/src/command-line/affected.ts]: should we inline it?

File: packages/workspace/src/core/file-utils.ts [hunk 6]
```diff
@@ -35,29 +44,30 @@ export function calculateFileChanges(
...
+            const atHead = readFileAtRevision(f, nxArgs.head);
+
+            try {
+              return jsonDiff(JSON.parse(atBase), JSON.parse(atHead));
+            } catch (e) {
+              return [WholeFileChange];
+            }
+          default:
+            return [WholeFileChange];
```
Comment (vsavkin) [inline packages/workspace/src/core/file-utils.ts]: This line shouldn't affect anything. Cause we never invoke this, right?

**Review by vsavkin — CHANGES_REQUESTED** [review]

**Comment by vsavkin** [comment]
Is the fix the try-catch around the JSON diff?

Can you also add a test somewhere to capture that?

File: packages/workspace/src/core/file-utils.ts (hunk 4 above)
Comment (FrozenPandaz) [reply packages/workspace/src/core/file-utils.ts]: We have to refactor the flow of processing the args a bit to do this. I'd rather do that separately.

File: packages/workspace/src/core/file-utils.ts (hunk 6 above)
Comment (FrozenPandaz) [reply packages/workspace/src/core/file-utils.ts]: It's not invoked right now.

File: packages/workspace/src/core/file-utils.ts (hunk 4 above)
Comment (vsavkin) [reply packages/workspace/src/core/file-utils.ts]: ok

File: packages/workspace/src/command-line/shared.ts [hunk 7]
```diff
@@ -1,11 +1,11 @@
 import { execSync } from 'child_process';
 import { output } from '../utils/output';
 import { createProjectGraph, ProjectGraphNode } from '../core/project-graph';
-import { NxArgs } from './utils';
 import { NxJson } from '../core/shared-interfaces';
 import { readWorkspaceJson, TEN_MEGABYTES } from '../core/file-utils';
+import * as yargs from 'yargs';
 
-export function printArgsWarning(options: NxArgs) {
+export function printArgsWarning(options: yargs.Arguments) {
```
Comment (vsavkin) [inline packages/workspace/src/command-line/shared.ts]: Why this change?

File: packages/workspace/src/command-line/shared.ts [hunk 8]
```diff
@@ -39,7 +39,7 @@ export function printArgsWarning(options: NxArgs) {
   }
 }
 
-export function parseFiles(options: NxArgs): { files: string[] } {
+export function parseFiles(options: yargs.Arguments): { files: string[] } {
```
Comment (vsavkin) [inline packages/workspace/src/command-line/shared.ts]: Why this change?

File: packages/workspace/src/core/file-utils.ts [hunk 9]
```diff
@@ -18,14 +19,29 @@ export interface FileData {
...
+
+export class WholeFileChange implements Change {
+  type = 'WholeFileChange';
+}
+
+export function isWholeFileChange(change: Change): change is WholeFileChange {
+  return change.type === 'WholeFileChange';
+}
+
+export interface FileChange<T extends Change = Change> extends FileData {
```
Comment (vsavkin) [inline packages/workspace/src/core/file-utils.ts]: File change appears twice (see line 26)

File: packages/workspace/src/command-line/format.ts [hunk 10]
```diff
@@ -57,7 +57,9 @@ export function format(command: 'check' | 'write', args: yargs.Arguments) {
   }
 }
 
-function getPatterns(args: NxArgs & { libsAndApps: boolean; _: string[] }) {
+function getPatterns(
+  args: NxArgs & { libsAndApps: boolean; _: string[] } & yargs.Arguments
```
Comment (vsavkin) [inline packages/workspace/src/command-line/format.ts]: why is yargs.Arguments required?

File: packages/workspace/src/command-line/shared.ts (hunk 8 above)
Comment (vsavkin) [reply packages/workspace/src/command-line/shared.ts]: We should only use the Arguments type when talking about executing tasks. The analysis part should not use Arguments.

File: packages/workspace/src/core/affected-project-graph/locators/implicit-json-changes.ts [hunk 11]
```diff
@@ -17,11 +21,20 @@ export function getImplicitlyTouchedProjectsByJsonChanges(
...
       for (const c of changes) {
-        const projects =
-          getTouchedProjects(c.path, implicitDependencies[f.file]) || [];
-        projects.forEach(p => touched.push(p));
+        if (isJsonChange(c)) {
+          const projects =
+            getTouchedProjects(c.path, implicitDependencies[f.file]) || [];
+          projects.forEach(p => touched.push(p));
+        } else {
+          walkJsonTree(implicitDependencies[f.file], [], (p, value) => {
```
Comment (vsavkin) [inline packages/workspace/src/core/affected-project-graph/locators/implicit-json-changes.ts]: what are you trying to do here?

I think this should be extracted into a function otherwise the if and else clauses operate on different levels of abstraction. One talks about touched projects, the other one recursively walks the tree

File: packages/workspace/src/core/affected-project-graph/locators/workspace-projects.ts [hunk 12]
```diff
@@ -18,6 +18,7 @@ export function getTouchedProjects(
 
 export function getImplicitlyTouchedProjects(
   workspaceJson: any,
+
```
Comment (vsavkin) [inline packages/workspace/src/core/affected-project-graph/locators/workspace-projects.ts]: Empty line.

File: packages/workspace/src/core/affected-project-graph/affected-project-graph.ts [hunk 13]
```diff
@@ -10,23 +16,35 @@ import { getImplicitlyTouchedProjectsByJsonChanges } from './locators/implicit-j
 import { AffectedProjectGraphContext } from './affected-project-graph-models';
 import { normalizeNxJson } from '../normalize-nx-json';
 
+export interface TouchedProjectLocator {
```
Comment (vsavkin) [inline packages/workspace/src/core/affected-project-graph/affected-project-graph.ts]: Affected-project-graph models might be a better place to have it

Comment (vsavkin) [reply packages/workspace/src/core/affected-project-graph/affected-project-graph.ts]: And since you introduced it, you can update all existing locators to use the interface

**Review by vsavkin — CHANGES_REQUESTED** [review]

File: packages/workspace/src/core/affected-project-graph/locators/implicit-json-changes.ts (hunk 11 above)
Comment (FrozenPandaz) [reply packages/workspace/src/core/affected-project-graph/locators/implicit-json-changes.ts]: This recursively finds projects affected by a JSON file.

```
implicitDependencies: {
  'package.json': {
    dependencies: {
      '@angular/core': ['proj1'],
    },
    devDependencies: ['proj2']
  }
}
```

Should touch `['proj1', 'proj2']` when touched.

Comment (vsavkin) [reply packages/workspace/src/core/affected-project-graph/locators/implicit-json-changes.ts]: Cool. Could you extract it into a helper function saying something like: `collectProjectsAffectedByAllChangesToPackageJson`
